File: java/src/Loaders.java
/**
 * Document Loaders — Java
 * Load TXT, HTML, Markdown, PDF, and DOCX into RagEngine.Document objects.
 * PDF: delegates to pdftotext (poppler-utils) — no Java PDF library CVE exposure.
 * DOCX: reads word/document.xml from ZIP — no extra dependency needed.
 *
 * apt-get install -y poppler-utils   (for PDF support)
 */

import java.io.*;
import java.nio.charset.StandardCharsets;
import java.nio.file.*;
import java.nio.file.attribute.BasicFileAttributes;
import java.util.*;
import java.util.regex.*;
import java.util.zip.*;
import javax.xml.parsers.*;
import org.w3c.dom.*;

public class Loaders {

    static final long MAX_FILE_BYTES = 50L * 1024 * 1024; // 50 MB
    static final Set<String> SUPPORTED = Set.of(".txt", ".md", ".markdown", ".html", ".htm", ".pdf", ".docx");

    static void checkSize(Path path) throws IOException {
        long size = Files.size(path);
        if (size > MAX_FILE_BYTES) {
            throw new IOException("File " + path + " is " + size + " bytes, exceeds " + MAX_FILE_BYTES + " byte limit");
        }
    }

    static String ext(Path path) {
        String name = path.getFileName().toString().toLowerCase();
        int dot = name.lastIndexOf('.');
        return dot >= 0 ? name.substring(dot) : "";
    }

    // ── TXT ───────────────────────────────────────────────────────────────────
    public static RagEngine.Document loadTxt(Path path) throws IOException {
        checkSize(path);
        String content = Files.readString(path, StandardCharsets.UTF_8);
        return new RagEngine.Document(content, Map.of("source", path.getFileName().toString(), "type", "txt"));
    }

    // ── Markdown ──────────────────────────────────────────────────────────────
    public static RagEngine.Document loadMarkdown(Path path) throws IOException {
        checkSize(path);
        String raw = Files.readString(path, StandardCharsets.UTF_8);
        // Strip Markdown syntax
        String text = raw
            .replaceAll("```[\\s\\S]*?```", " ")
            .replaceAll("`[^`]+`", " ")
            .replaceAll("#{1,6}\\s+", "")
            .replaceAll("\\*{1,2}([^*]+)\\*{1,2}", "$1")
            .replaceAll("!?\\[([^\\]]+)\\]\\([^)]+\\)", "$1")
            .replaceAll("\\s+", " ")
            .trim();
        return new RagEngine.Document(text, Map.of("source", path.getFileName().toString(), "type", "markdown"));
    }

    // ── HTML ──────────────────────────────────────────────────────────────────
    public static RagEngine.Document loadHtml(Path path) throws IOException {
        checkSize(path);
        String raw = Files.readString(path, StandardCharsets.UTF_8);
        // Extract title
        Matcher tm = Pattern.compile("(?i)<title[^>]*>([^<]*)</title>").matcher(raw);
        String title = tm.find() ? tm.group(1).trim() : "";
        // Strip script/style blocks then all tags
        String clean = raw
            .replaceAll("(?is)<script[^>]*>.*?</script>", " ")
            .replaceAll("(?is)<style[^>]*>.*?</style>", " ")
            .replaceAll("<[^>]+>", " ")
            .replaceAll("\\s+", " ")
            .trim();
        Map<String, String> meta = new LinkedHashMap<>();
        meta.put("source", path.getFileName().toString());
        meta.put("type", "html");
        meta.put("title", title);
        return new RagEngine.Document(clean, meta);
    }

    // ── PDF ───────────────────────────────────────────────────────────────────
    public static RagEngine.Document loadPdf(Path path) throws IOException, InterruptedException {
        checkSize(path);
        ProcessBuilder pb = new ProcessBuilder("pdftotext", "-enc", "UTF-8", path.toString(), "-");
        pb.redirectErrorStream(true);
        Process proc = pb.start();
        String output;
        try (InputStream is = proc.getInputStream()) {
            output = new String(is.readAllBytes(), StandardCharsets.UTF_8).trim();
        }
        int exit = proc.waitFor();
        if (exit != 0) {
            throw new IOException("pdftotext failed (exit " + exit + ") — install poppler-utils");
        }
        return new RagEngine.Document(output, Map.of("source", path.getFileName().toString(), "type", "pdf"));
    }

    // ── DOCX ──────────────────────────────────────────────────────────────────
    public static RagEngine.Document loadDocx(Path path) throws Exception {
        checkSize(path);
        try (ZipFile zip = new ZipFile(path.toFile())) {
            ZipEntry entry = zip.getEntry("word/document.xml");
            if (entry == null) {
                throw new IOException("word/document.xml not found in " + path);
            }
            try (InputStream is = zip.getInputStream(entry)) {
                DocumentBuilderFactory factory = DocumentBuilderFactory.newInstance();
                // Disable external entity processing (XXE protection)
                factory.setFeature("http://apache.org/xml/features/disallow-doctype-decl", true);
                factory.setFeature("http://xml.org/sax/features/external-general-entities", false);
                factory.setFeature("http://xml.org/sax/features/external-parameter-entities", false);
                factory.setXIncludeAware(false);
                factory.setExpandEntityReferences(false);

                org.w3c.dom.Document xml = factory.newDocumentBuilder().parse(is);
                NodeList nodes = xml.getElementsByTagNameNS("*", "t");
                StringBuilder sb = new StringBuilder();
                for (int i = 0; i < nodes.getLength(); i++) {
                    String text = nodes.item(i).getTextContent().trim();
                    if (!text.isEmpty()) {
                        if (sb.length() > 0) sb.append(" ");
                        sb.append(text);
                    }
                }
                return new RagEngine.Document(sb.toString(), Map.of("source", path.getFileName().toString(), "type", "docx"));
            }
        }
    }

    // ── Auto-dispatch ─────────────────────────────────────────────────────────
    public static RagEngine.Document loadDocument(Path path) throws Exception {
        String e = ext(path);
        return switch (e) {
            case ".txt"                     -> loadTxt(path);
            case ".md", ".markdown"         -> loadMarkdown(path);
            case ".html", ".htm"            -> loadHtml(path);
            case ".pdf"                     -> loadPdf(path);
            case ".docx"                    -> loadDocx(path);
            default -> throw new IOException("Unsupported extension '" + e + "'");
        };
    }

    public static List<RagEngine.Document> loadDirectory(Path dir, boolean recursive) throws Exception {
        List<RagEngine.Document> docs = new ArrayList<>();
        int depth = recursive ? Integer.MAX_VALUE : 1;
        Files.walkFileTree(dir, Set.of(), depth, new SimpleFileVisitor<>() {
            @Override
            public FileVisitResult visitFile(Path file, BasicFileAttributes attrs) {
                if (SUPPORTED.contains(ext(file))) {
                    try {
                        docs.add(loadDocument(file));
                    } catch (Exception e) {
                        System.err.println("[loaders] skipping " + file + ": " + e.getMessage());
                    }
                }
                return FileVisitResult.CONTINUE;
            }
        });
        return docs;
    }
}

File: go/src/loaders.go
// Document Loaders — Go
// Load TXT, HTML, Markdown, PDF, and DOCX into Document objects.
// PDF: uses pdftotext (poppler-utils, no Go CVEs). DOCX: parses word/document.xml directly.
//
// apt-get install -y poppler-utils   (for PDF support)
// go run ./src/... load <path>

package main

import (
	"archive/zip"
	"encoding/xml"
	"fmt"
	"io"
	"os"
	"os/exec"
	"path/filepath"
	"regexp"
	"strings"
)

const goMaxFileBytes = 50 * 1024 * 1024 // 50 MB

var goSupportedExts = map[string]bool{
	".txt": true, ".md": true, ".markdown": true,
	".html": true, ".htm": true, ".pdf": true, ".docx": true,
}

func checkGoFileSize(path string) error {
	info, err := os.Stat(path)
	if err != nil {
		return err
	}
	if info.Size() > goMaxFileBytes {
		return fmt.Errorf("file %s is %d bytes, exceeds %d byte limit", path, info.Size(), goMaxFileBytes)
	}
	return nil
}

func loadTxtGo(path string) (*Document, error) {
	if err := checkGoFileSize(path); err != nil {
		return nil, err
	}
	data, err := os.ReadFile(path)
	if err != nil {
		return nil, err
	}
	return &Document{
		Content:  string(data),
		Metadata: map[string]string{"source": filepath.Base(path), "type": "txt"},
	}, nil
}

var htmlTagRe  = regexp.MustCompile(`(?s)<script[^>]*>.*?</script>|<style[^>]*>.*?</style>|<[^>]+>`)
var htmlTitleRe = regexp.MustCompile(`(?i)<title[^>]*>([^<]*)</title>`)
var wsRe        = regexp.MustCompile(`\s+`)

func loadHTMLGo(path string) (*Document, error) {
	if err := checkGoFileSize(path); err != nil {
		return nil, err
	}
	data, err := os.ReadFile(path)
	if err != nil {
		return nil, err
	}
	raw := string(data)
	title := ""
	if m := htmlTitleRe.FindStringSubmatch(raw); len(m) > 1 {
		title = strings.TrimSpace(m[1])
	}
	clean := htmlTagRe.ReplaceAllString(raw, " ")
	clean  = strings.TrimSpace(wsRe.ReplaceAllString(clean, " "))
	return &Document{
		Content:  clean,
		Metadata: map[string]string{"source": filepath.Base(path), "type": "html", "title": title},
	}, nil
}

func loadMarkdownGo(path string) (*Document, error) {
	if err := checkGoFileSize(path); err != nil {
		return nil, err
	}
	data, err := os.ReadFile(path)
	if err != nil {
		return nil, err
	}
	// Strip Markdown syntax (headings, bold, italic, links, code fences)
	text := string(data)
	text  = regexp.MustCompile("```[\\s\\S]*?```").ReplaceAllString(text, " ")
	text  = regexp.MustCompile("`[^`]+`").ReplaceAllString(text, " ")
	text  = regexp.MustCompile(`#{1,6}\s+`).ReplaceAllString(text, "")
	text  = regexp.MustCompile(`\*{1,2}([^*]+)\*{1,2}`).ReplaceAllString(text, "$1")
	text  = regexp.MustCompile(`\[([^\]]+)\]\([^)]+\)`).ReplaceAllString(text, "$1")
	text  = strings.TrimSpace(wsRe.ReplaceAllString(text, " "))
	return &Document{
		Content:  text,
		Metadata: map[string]string{"source": filepath.Base(path), "type": "markdown"},
	}, nil
}

func loadPDFGo(path string) (*Document, error) {
	if err := checkGoFileSize(path); err != nil {
		return nil, err
	}
	// Use pdftotext (poppler-utils) — no Go PDF library has a clean CVE record at this scope
	out, err := exec.Command("pdftotext", "-enc", "UTF-8", path, "-").Output()
	if err != nil {
		return nil, fmt.Errorf("pdftotext failed (install poppler-utils): %w", err)
	}
	content := strings.TrimSpace(string(out))
	return &Document{
		Content:  content,
		Metadata: map[string]string{"source": filepath.Base(path), "type": "pdf"},
	}, nil
}

// DOCX: a .docx is a ZIP containing word/document.xml — no external library needed.
type docxBody struct {
	Paragraphs []docxPara `xml:"body>p"`
}

type docxPara struct {
	Runs []docxRun `xml:"r"`
}

type docxRun struct {
	Text string `xml:"t"`
}

func loadDOCXGo(path string) (*Document, error) {
	if err := checkGoFileSize(path); err != nil {
		return nil, err
	}
	r, err := zip.OpenReader(path)
	if err != nil {
		return nil, fmt.Errorf("failed to open docx: %w", err)
	}
	defer r.Close()

	for _, f := range r.File {
		if f.Name != "word/document.xml" {
			continue
		}
		rc, err := f.Open()
		if err != nil {
			return nil, err
		}
		defer rc.Close()
		data, err := io.ReadAll(io.LimitReader(rc, goMaxFileBytes))
		if err != nil {
			return nil, err
		}
		var body docxBody
		if err := xml.Unmarshal(data, &body); err != nil {
			return nil, fmt.Errorf("failed to parse docx XML: %w", err)
		}
		var parts []string
		for _, para := range body.Paragraphs {
			var sb strings.Builder
			for _, run := range para.Runs {
				sb.WriteString(run.Text)
			}
			if text := strings.TrimSpace(sb.String()); text != "" {
				parts = append(parts, text)
			}
		}
		content := strings.Join(parts, "\n")
		return &Document{
			Content:  content,
			Metadata: map[string]string{"source": filepath.Base(path), "type": "docx"},
		}, nil
	}
	return nil, fmt.Errorf("word/document.xml not found in %s", path)
}

// ── Auto-dispatch ──────────────────────────────────────────────────────────────
func LoadDocument(path string) (*Document, error) {
	ext := strings.ToLower(filepath.Ext(path))
	switch ext {
	case ".txt":
		return loadTxtGo(path)
	case ".md", ".markdown":
		return loadMarkdownGo(path)
	case ".html", ".htm":
		return loadHTMLGo(path)
	case ".pdf":
		return loadPDFGo(path)
	case ".docx":
		return loadDOCXGo(path)
	default:
		return nil, fmt.Errorf("unsupported extension '%s'", ext)
	}
}

func LoadDirectory(dir string, recursive bool) ([]*Document, error) {
	var docs []*Document
	walkFn := func(path string, info os.FileInfo, err error) error {
		if err != nil {
			return nil
		}
		if info.IsDir() {
			if !recursive && path != dir {
				return filepath.SkipDir
			}
			return nil
		}
		ext := strings.ToLower(filepath.Ext(path))
		if !goSupportedExts[ext] {
			return nil
		}
		doc, err := LoadDocument(path)
		if err != nil {
			fmt.Fprintf(os.Stderr, "[loaders] skipping %s: %v\n", path, err)
			return nil
		}
		docs = append(docs, doc)
		return nil
	}
	if err := filepath.Walk(dir, walkFn); err != nil {
		return nil, err
	}
	return docs, nil
}
